interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,310 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/utils"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/systemd"
+)
+
+const mountControlSummary = `allows creating transient and persistent mounts`
+
+const mountControlBaseDeclarationSlots = `
+  mount-control:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
+    deny-connection: true
+`
+
+const mountControlConnectedPlugSecComp = `
+# Description: Allow mount and umount syscall access.
+mount
+umount
+umount2
+`
+
+var allowedMountOptions = []string{
+	"async",
+	"atime",
+	"bind",
+	"diratime",
+	"dirsync",
+	"iversion",
+	"lazytime",
+	"mand",
+	"nofail",
+	"noiversion",
+	"nomand",
+	"noatime",
+	"nodev",
+	"nodiratime",
+	"noexec",
+	"nolazytime",
+	"norelatime",
+	"nosuid",
+	"nostrictatime",
+	"nouser",
+	"relatime",
+	"strictatime",
+	"sync",
+	"ro",
+	"rw",
+}
+
+// mountControlInterface allows creating transient and persistent mounts
+type mountControlInterface struct {
+	commonInterface
+}
+
+// TODO: review the regexp
+// We prohibit expressions containing a comma, as that is used by AppArmor to
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
+
+type MountInfo struct {
+	what       string
+	where      string
+	persistent bool
+	types      []string
+	options    []string
+}
+
+func parseStringList(mountEntry map[string]interface{}, fieldName string) ([]string, error) {
+	var list []string
+	value, ok := mountEntry[fieldName]
+	if ok {
+		interfaceList, ok := value.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf(`mount-control "mount::%s" must be an array of strings (got %q)`, fieldName, value)
```

Comment:
we don't use "::" as separator like this,  we should probably either use "." or be vervose and use "mount entry "%s"

Reply:
all the validation error cases need test coverage

Reply:
Done.